### PR TITLE
Fixes #29135 - Move external links to a single controller

### PR DESCRIPTION
**Pull request by ShimShtein**

### CODE REVIEW

File: app/controllers/links_controller.rb
Comment (ekohl) [inline app/controllers/links_controller.rb]:
```suggestion
      'https://projects.theforeman.org/projects/foreman/issues'
```

File: app/controllers/links_controller.rb
Comment (ekohl) [inline app/controllers/links_controller.rb]: This redirects
```suggestion
      'https://freenode.net'
```

**Review by ekohl — CHANGES_REQUESTED** [review]

**Comment by ShimShtein** [comment]
@ekohl, done.

**Comment by timogoebel** [comment]
Does this "override" the referrer to not expose any sensitive information like hostnames in the URL?

File: app/controllers/links_controller.rb
Comment (ekohl) [inline app/controllers/links_controller.rb:1]: Is this a controller instead of a service because of Javascript?

Comment (ShimShtein) [reply app/controllers/links_controller.rb:1]: Yes, the idea is that once we have this controller, we can generate links on the server, classic JS, angular or react. All of them will go through a single controller that will be responsible for bookkeeping of all the external links.

**Comment by ShimShtein** [comment]
> Does this "override" the referrer to not expose any sensitive information like hostnames in the URL?

@timogoebel, I have left the state as it was, so if the links were opened in new tab, or whatever, they still will be opened the same way.
The links that are passed here do not contain any sensitive information, only chapter names in documentation urls. Do you think we need to add such features when we access our documentation?

**Review by lzap — CHANGES_REQUESTED** [review]
I like the idea, but this needs to go much further in order to improve this. Let me explain.

We are currently in transition to new generation of docs, so actually we now have three sets of documentation links:

* https://theforeman.org/manuals
* https://docs.theforeman.org/web/
* https://access.redhat.com/documentation/en-us/red_hat_satellite

Each one of them have different anchors. The goal is to unify this but we are not there yet and it will be a longer process. Therefore this PR should make it better for the future and get ready for coming changes in this regard. I suggest:

1. Get rid of "4.4.3SomeAnchor" naming convention. It's bad from the day one, chapter numbers should have never been there. I know this is disturbing change, but there are not many places in the codebase for replace. Let's come with some good naming convention like "host_create_form" or "subnet_edit_page". Something that could be eventually also used as anchors in the documentation itself (a simple string without spaces).

1. Let's create a single place with all the links. Currently the documentation links are spread across the codebase, we should have a single place (a Ruby file or YAML) with all the links. This way we have a nice overview of the links but more importantly:

1. Let's create a rake task that will perform HTTP HEAD calls and validations of all links. We suffer from invalid documentation links very often. This could easily stop it if we add this check into our nightly pipeline or ask QEs to perform this test on each release.

1. (optional) Let's take out upstream/downstream documentation URL link handling from branding plugin into core. I don't think that it's major issue to have this exposed and have all the link in a single place. It's actually needed if we want to unify links across downstream/upstream. I believe we should take the opportunity and create a single place for links in foreman core, branding plugin would only turn on some switch to use downstream ones. Alternatively, the plugin could override the configuration file if needed.

**Comment by lzap** [comment]
Also I am playing around with idea of creating an appendix or even a separate document with a single purpose: to be a landing page for links in Foreman. We would simply create anchors like "host_edit_page" with links to various places in our documentation. Advantage of this approach is that a single link from Foreman can actually branch off into multiple places within our documentation. Also it's easier for the docs team not to break HTML anchors if all of them are on a single place.

The problem is that the documentation team moves chapters around and links eventually break unnoticed. This "appendix" would create a single place where docs team would know these anchors must not be changed. I believe this could strongly increase this feature and we could even start adding more and more links which is a good way of improving Foreman user experience. In the documentation, we do have a link validator that makes sure no links are invalid. Therefore once we make sure links in Foreman are correct as well, the whole loop for broken links is closed.

Example report from last week: https://bugzilla.redhat.com/show_bug.cgi?id=1804651 - Smart Proxy create form link lands on page that only explain concept of Smart Proxy downstream but does not provide instructions about create entry procedure.

This is very much forward looking idea but it goes well with what we do upstream with @melcorr, what you think?

**Comment by melcorr** [comment]
Hi @lzap I think I understand what you mean, and I am OK with that as an alternative way for users to orient themselves within our docs.

**Comment by ShimShtein** [comment]
> Get rid of "4.4.3SomeAnchor" naming convention. It's bad from the day one, chapter numbers should have never been there. I know this is disturbing change, but there are not many places in the codebase for replace. Let's come with some good naming convention like "host_create_form" or "subnet_edit_page". Something that could be eventually also used as anchors in the documentation itself (a simple string without spaces).

I love the idea, since it makes more sense to name the anchors by the source of the link, especially when we have a controller that may be the place that translates between link sources to documentation targets.
My only question is do we want to change the convention in the same PR? Theoretically it doesn't have to be.

> Let's create a single place with all the links. Currently the documentation links are spread across the codebase, we should have a single place (a Ruby file or YAML) with all the links. 

Links controller is exactly that - it is a single place where all links will be directed to. If we go with the idea from (1) to translate every link, we will have a single dictionary where a key is the source and the value is the destination URL. This dictionary is the overview you are looking for.

> Let's create a rake task that will perform HTTP HEAD calls and validations of all links. We suffer from invalid documentation links very often.

Since downstream I am doing something very similar to this approach, I do have a fully automated test that pings all available URLs to ensure they are alive.

The only issue is with making people actually use this controller, especially on client side. There is no way to enforce it and everyone still can hardcode external links.

**Comment by timogoebel** [comment]
> The links that are passed here do not contain any sensitive information, only chapter names in documentation urls. Do you think we need to add such features when we access our documentation?

I was not referring to the documentation URLs (they are harmless) but the page that links to these URLs (that could be `https://foreman.secret.cia.area51.example.com/hosts/my-very-secret-alien-hostname.secure.site.cia.gov.example.com`) and end up in the access logs of the documentation's server.

**Comment by ShimShtein** [comment]
> I was not referring to the documentation URLs (they are harmless) but the page that links to these URLs (that could be https://foreman.secret.cia.area51.example.com/hosts/my-very-secret-alien-hostname.secure.site.cia.gov.example.com) and end up in the access logs of the documentation's server.

@timogoebel, good point. I don't think I make things worse than they already are. Although I can be mistaken. Let's see, in the previous state the user clicks directly on documentation url, so you have the referrer set to your sensitive host. After this PR, the link will access local foreman and get 302 response. Then the browser will follow the redirect and get to our docs server. From googling, it looks like the referrer will remain the same as without the redirect - hence still the sensitive host. On the other hand, instead of simple redirect, we can use some referrer hiding technique here. Do I smell a plugin here :wink: ?

**Comment by ekohl** [comment]
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referrer-Policy suggests you could send a header in this controller and solve it. The `secure_headers` gem can also set this globally.

**Review by lzap — DISMISSED** [review]
I agree that we should keep PRs isolated, if I want to pull the dictionary of links from our branding plugin upstream that should be a separate PR. The same goes for the referer problem, only if this patch would be introducing such a security concern I'd block merge but that's not the case. I am going to test this and give it a go.

The referer problem is a good one, I have created an issue for that: https://projects.theforeman.org/issues/29206

File: app/views/about/index.html.erb
Comment (ekohl) [inline app/views/about/index.html.erb:130]: This is an odd case. It linked to Redmine in general with the label wiki. Since there's an issue tracker below, this looks correct.

File: webpack/assets/javascripts/react_app/routes/Audits/constants.js
Comment (ekohl) [inline webpack/assets/javascripts/react_app/routes/Audits/constants.js]: Perhaps the method should take an additional argument:
```suggestion
  getManualURL(version, '4.1.4Auditing');
```

File: webpack/assets/javascripts/react_app/common/helpers.js
Comment (ekohl) [inline webpack/assets/javascripts/react_app/common/helpers.js]: There's a `foremanUrl` helper which actually gets the correct URLs to Foreman since they're not guaranteed to be at `/`.

**Review by ekohl — CHANGES_REQUESTED** [review]
You might want to get one more: https://github.com/theforeman/foreman/blob/f75b27377e82fd9760a67764b64cbc17d66ef3c9/app/views/config_reports/welcome.html.erb#L11

File: app/views/about/index.html.erb
Comment (ekohl) [inline app/views/about/index.html.erb:133]: I'd name this `freenode` since the hostname (`irc.freenode.net`) is hardcoded in the translation string as well. Feels to name it chat given the header is also `IRC`. Sure, it can be changed via translations but then you have yet another thing to translate very illogically.

```suggestion
      <p><%= (_("You can find The Foreman on the %{freenode} (irc.freenode.net) network.  For general support, please visit #theforeman and for development specific related chat, please visit #theforeman-dev.") % {:freenode => link_to("freenode", external_link_path(type: 'freenode'), :rel => "external")}).html_safe  %></p>
```

**Review by ekohl — COMMENTED** [review]
Why isn't the audits page snapshot failing? https://github.com/theforeman/foreman/blob/44e19b3546c6496b8f6de2ec3ab663db8eb149ab/webpack/assets/javascripts/react_app/routes/Audits/AuditsPage/__tests__/__snapshots__/AuditsPage.test.js.snap#L39 shouldn't be the same anymore, right?

File: app/views/about/index.html.erb
Comment (ShimShtein) [reply app/views/about/index.html.erb:133]: I think the link should remain general. As for the text - there is a way to override the translated text, so this way the text and the link will be redirected to a different location.

**Comment by lzap** [comment]
Can you rebase please? I will test and merge. @ekohl your review seemed to be incorporated, can you re-check please or mark it as ok?

File: app/controllers/links_controller.rb
Comment (ekohl) [inline app/controllers/links_controller.rb:31]: This particular page isn't really showing anything anymore and just redirecting to the quickstart guide. It doesn't respect the version so you end up at the latest stable quickstart. This was probably also an issue before but probably something to look at. Given you're just moving code, it's fine to do it in a later PR.

File: webpack/assets/javascripts/react_app/routes/Audits/constants.js
Comment (ekohl) [inline webpack/assets/javascripts/react_app/routes/Audits/constants.js]: General javascript question: is there a reason to prefer `'` here given there's no variable interpolation going on?

**Review by ekohl — COMMENTED** [review]
Overall :+1: once the merge conflict has been solved.

**Comment by ShimShtein** [comment]
Travis failed on coveralls

**Comment by lzap** [comment]
Well this looks like a failure, can you investigate if that's related?

```
console.error node_modules/jest-prop-type-error/index.js:8

    Error: Can't find a registered interval process for: NOTIFICATIONS

        at unregisteredIntervalException (/home/travis/build/theforeman/foreman/webpack/assets/javascripts/react_app/redux/middlewares/IntervalMiddleware/IntervalHelpers.js:7:3)

        at /home/travis/build/theforeman/foreman/webpack/assets/javascripts/react_app/redux/middlewares/IntervalMiddleware/IntervalMiddleware.js:37:13

        at /home/travis/build/theforeman/foreman/node_modules/redux-thunk/lib/index.js:14:16

        at /home/travis/build/theforeman/foreman/node_modules/redux-thunk/lib/index.js:14:16

        at stopNotificationsPolling (/home/travis/build/theforeman/foreman/node_modules/redux/lib/redux.js:483:12)

        at notificationContainer.componentDidUpdate (/home/travis/build/theforeman/foreman/webpack/assets/javascripts/react_app/components/notifications/index.js:39:7)

        at commitLifeCycles (/home/travis/build/theforeman/foreman/node_modules/react-dom/cjs/react-dom.development.js:19835:22)

        at commitLayoutEffects (/home/travis/build/theforeman/foreman/node_modules/react-dom/cjs/react-dom.development.js:22803:7)
```

File: webpack/assets/javascripts/react_app/routes/Audits/constants.js
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/routes/Audits/constants.js]: I agree, we can use regular single qoutes
and it doesn't need to be a function since we dropped the version param, right?

Comment (ShimShtein) [reply webpack/assets/javascripts/react_app/routes/Audits/constants.js]: The function will make the evaluation in runtime (which I think is still desirable).

Comment (ekohl) [reply webpack/assets/javascripts/react_app/routes/Audits/constants.js]: `getManualURL` itself is still dynamic because `getForemanUrl` is still dynamic I think. Not entirely sure, but this must be resolved on the client at runtime, rather than during build time.

**Review by ekohl — DISMISSED** [review]

**Comment by ShimShtein** [comment]
@lzap, now everything passes. Can you do another quick round?

**Comment by lzap** [comment]
Sorry for the delay, currently testing this. Can you raise a BZ for related downstream changes so we won't forget to do them (if there are any).

**Review by lzap — APPROVED** [review]
I clicked on all of them except VMWare link which should work too. Thanks.

**Pull request merged**
